extensions/amp-analytics/0.1/amp-analytics-3p-message-queue.js

```diff
@@ -0,0 +1,187 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {dev} from '../../../src/log';
+import {AMP_ANALYTICS_3P_MESSAGE_TYPE} from '../../../src/3p-analytics-common';
+import {throttle} from '../../../src/utils/rate-limit';
+
+/** @private @const {string} */
+const TAG_ = 'amp-analytics.CrossDomainIframeMessageQueue';
+
+/** @private @const {number} */
+const MESSAGE_THROTTLE_TIME_ = 100;
+
+/** @private @const {number} */
+const MAX_QUEUE_SIZE_ = 100;
+
+/**
+ * @visibleForTesting
+ * @abstract
+ */
+class AbstractAmpAnalytics3pMessageQueue {
+  /**
+   * Constructor
+   * @param {!Window} win The window element
+   * @param {!../../../3p/iframe-messaging-client.IframeMessagingClient}
+   *   iframeMessagingClient Facilitates cross-frame communication
+   * @param {!string} messageType The type that will be used to
+   * contain the queued data (see 3p-analytics-common)
+   */
+  constructor(win, iframeMessagingClient, messageType) {
+    /** @private {boolean} */
+    this.isReady_ = false;
+
+    /** @private {!Window} */
+    this.win_ = win;
+
+    /** @private {!../../../3p/iframe-messaging-client.IframeMessagingClient} */
+    this.iframeMessagingClient_ = iframeMessagingClient;
+
+    /** @private {string} */
+    this.messageType_ = messageType;
+
+    /** @private {!Object<string,!string|!Array<string>>} */
+    this.creativeToPendingMessages_ = {};
+
+    /** @private {function(...*)} */
+    this.throttledFlushQueue_ = throttle(this.win_, () => {
+      this.flushQueue_();
+    }, MESSAGE_THROTTLE_TIME_);
+  }
+
+  /**
+   * Indicate that a cross-domain frame is ready to receive messages, and
+   * send all messages that were previously queued for it.
+   */
+  setIsReady() {
+    this.isReady_ = true;
+    this.throttledFlushQueue_();
+  }
+
+  /**
+   * Send queued data (if there is any) to a cross-domain iframe
+   * This should never be called directly. Use this.throttledFlushQueue_()
+   * instead.
+   * @private
+   */
+  flushQueue_() {
+    if (this.isReady_ && Object.keys(this.creativeToPendingMessages_).length) {
+      const jsonMsg = /** @type {JsonObject} */ (this.buildMessage_());
+      this.iframeMessagingClient_./*OK*/sendMessage(this.messageType_, jsonMsg);
+      this.creativeToPendingMessages_ = {};
+    }
+  }
+
+  /**
+   * @return {../../../src/3p-analytics-common.AmpAnalytics3pNewCreative|
+   *          ../../../src/3p-analytics-common.AmpAnalytics3pEvent}
+   * @abstract
+   */
+  buildMessage_() {
```

Comment:
Surprisingly, Travis will fail if it gets a warning. So, taking this throw back out :(
https://travis-ci.org/ampproject/amphtml/jobs/245496442